### ISSUE TITLE
Check module compatibility on update

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/wordpoints#330]
Since this is not a backward-compatible release, we should check the compatibility of all active modules when we update, and deactivate any that aren't compatible.

**Comment by JDGrimes** [comment]
## Multisite Considerations

I've been wondering what the best way would be to do this on multisite. We never need to check modules that aren't active, because they'll be checked for errors when they are activated. On a non-multisite we have a single list of the active modules, which makes it easy. But on multisite we have network-active modules and the list of modules which are active for each site. I was thinking that it would be too hard to get the list for every site, so maybe we should just check all of the modules. However, we already have the code in place to loop through all of the sites on update anyway. We should just do that, but keep a list of which module's have already been checked so we don't check the same modules twice on different sites.

This will mean that update could take a few minutes, but there won't be more than two or three modules installed on most sites. Also, we can run the compatibility checks for all of the modules on a site in a single request. We'd only have to break it up into multiple requests if some of the modules had errors.

**Comment by JDGrimes** [comment]
One problem with using the un/installer is that the updates aren't run until after modules are loaded. So we would need to run it earlier or skip loading the modules if the update hasn't run. We'll still be wanting to load the modules, so maybe we should run this earlier.

**Comment by JDGrimes** [comment]
We need to decide how aggressive we want to be. Do we want to do an error scrape like `wordpoints_activate_module()` where the module is loaded after WordPress's normal hooks have already run? But that will only check code which isn't hooked to an action. And that isn't very useful, because almost all code should be hooked to an action anyway. But if we fully load WordPress each time, we'll have to run the check for each module individually. Otherwise, we don't have any way to tell which module caused the error, if one occurs. Although we could probably attempt running them all together and then use an algorithm to identify the troublemaker(s). We could also just deactivate them all if we find an error, and ask the user to reactivate them one by one. But that would be problematic on multisite, because the admin would have to visit a bunch of different sites and reactivate the modules. It's also true that reactivation wouldn't solve the problem, because that only does the simple error check, which might not catch the issue.

Another thing to consider is that we'll need to implement some kind of maintenance mode while the modules are being checked. But I don't think we'll be able to use WordPress's built-in maintenance mode feature, because we have to be able to make requests back to the site.

**Comment by JDGrimes** [comment]
Also, if we go for a full-load check, should it be performed on the front end, or the admin, or both?

I guess we need to decide exactly what it is we are trying to do here. We can't possibly detect every broken thing, or every fatal error. The main thing we want to avoid is outright killing a user's site, white-screen-of-death style. The most egregious problem is when this happens on the administration panels, making it impossible for the user to fix the problem without FTP access or the like. 

Running the basic error check is not sufficient to detect that no WSoD will result from running the module with WordPoints 2.0. It is not sufficient to ensure that the module won't WSoD the administration panels. This is the biggest concern. So, if we want to do this right, we'll need to do the full-load check, at least on the admin panels.

So, let's start with that, and we can decide to check the front-end too later if we want to.

**Comment by JDGrimes** [comment]
What about if WordPoints is updated in a bulk-update with other plugins. Will that cause maintenance mode to be used and conflict with our custom maintenance mode?

**Comment by JDGrimes** [comment]
I think that we can hack WordPress's maintenance mode code by putting some logic into the `.maintenance` file. That logic would then set `$upgrading` as needed to either allow the request to go through (if `$upgrading` is more than 10 minutes old timestamp) or show the maintenance message.

**Comment by JDGrimes** [comment]
> What about if WordPoints is updated in a bulk-update with other plugins. Will that cause maintenance mode to be used and conflict with our custom maintenance mode?

This won't be a problem. There will be no requests hitting the site while maintenance mode is enabled, so our updates won't be triggered until after it has ended.

**Comment by JDGrimes** [comment]
I think the best way will be to run once with all of the modules. If there is a failure, we'll run through each of them one-by-one. When 3.0 is released we can change things if we think that we need to.

**Comment by JDGrimes** [comment]
What should we do when we can't enable maintenance mode? This could happen because of filesystem issues. In that case, I think we should attempt the compatibility check anyway. I see no reason why it shouldn't work.

**Comment by JDGrimes** [comment]
I've just realized that the error scrape that occurs when a module fails to activate is performed using purely redirects and and iframe. However, I was planning to use `wp_remote_*()` to make the error check requests. I not sure whether this will be an issue or not.

**Comment by JDGrimes** [comment]
OK, I've run into a snag. We can't load the modules page because we are making the request unauthenticated, and therefore do not have the permissions to view the modules admin screen. We can either try to hack our way around this, or use a publicly accessible page instead. I'm going to see how well it would work with Ajax.

**Comment by JDGrimes** [comment]
I've been able to get it working by using `admin-ajax.php` and loading the modules screen in the Ajax callback.

**Comment by JDGrimes** [comment]
I guess this is done. I've spun off #338.

**Comment by JDGrimes** [comment]
I missed one thing here. When a module has already been checked, we don't check it again. But if it was incompatible we still need to deactivate it.

**Comment by JDGrimes** [comment]
Another thing that we need to do is notify the user about any modules that were deactivated.

**Comment by JDGrimes** [comment]
> Another thing that we need to do is notify the user about any modules that were deactivated.

On multisite, this poses an interesting dilemma. We want the administrator to know immediately which modules were deactivated, so that they can take appropriate action. However, they'll also want to be able to reactivate or replace them everywhere that they were active. So should we store a global list, or a per-site list, or both?

Maybe we even want to consider a different approach than deactivating, like temporarily disabling them somehow instead. But that seems like overkill. 

I guess what I'm getting at is that ideally in this situation the admin would have the option to roll back to the previous version and reactivate all of the deactivated modules. But we'd need to also roll back any database changes, or else skip the database changes. But then the database might be incompatible with the code, and so the plugin would be broken. I don't think that there are any breaking database changes in the update from 1.10 to 2.0, but what if the update is from 1.6 to 2.0, for example? There would be breaking database changes then. Also, when we rolled back the code, would we roll back to 1.10, or to the previous version?

All things considered, that's more than we have time to figure out and code before release.

In conclusion, I think that we should store a global list of all deactivated modules to show the network administrators, and a list of modules deactivated on each site to show the site admins.

**Comment by JDGrimes** [comment]
OK, I think this is really done this time.
